### PR TITLE
Implement equipment card view

### PR DESCRIPTION
**Pull request by ywsang**
## Overview

Implemented the new cells for be used for the "EQUIPMENT" section in the new Facilities design. 

## Changes Made

- Created `EquipmentListCell` and `EquipmentListItemCell` for the collection view (to be used as a cell in the dropdown) and the individual cells in the collection view
- Added fields to the all gyms query in order to retrieve facility information 
- Added models to represent the new FacilityType and EquipmentType in the GraphQL schema

## Test Coverage

Ran on simulator embedded in `GymDetailViewController`, video in #uplift-ios.

## Related PRs or Issues (delete if not applicable)

Addresses #166 

## Screenshots (delete if not applicable)

Video in #uplift-ios

### CODE REVIEW

File: Uplift/Controllers/GymDetailViewController.swift [hunk 1]
```diff
@@ -49,6 +50,10 @@ class GymDetailViewController: UIViewController {
         super.init(nibName: nil, bundle: nil)
         self.gymDetail = GymDetail(gym: gym)
 
+        for facility in gymDetail.facilities where facility.name == "Fitness Center" {
```
Comment (kevinchan159) [inline Uplift/Controllers/GymDetailViewController.swift]: is "Fitness Center" used in more than one file? I feel like it probably is and we should make it a global constant

File: Uplift/Controllers/GymDetailViewController.swift [hunk 2]
```diff
@@ -282,6 +287,31 @@ func getHoursHeight() -> CGFloat {
     }
 }
 
+extension GymDetailViewController {
+    func categorizeEquipment(equipmentList: [Equipment]) -> [EquipmentCategory] {
+        var equipmentDictionary: [String: [Equipment]] = [:]
+        let sorted = equipmentList.sorted(by: { $0.equipmentType < $1.equipmentType })
```
Comment (kevinchan159) [inline Uplift/Controllers/GymDetailViewController.swift]: `sortedEquipmentList`

File: Uplift/Controllers/GymDetailViewController.swift [hunk 3]
```diff
@@ -282,6 +287,31 @@ func getHoursHeight() -> CGFloat {
...
+        var category = sorted[0].equipmentType
```
Comment (kevinchan159) [inline Uplift/Controllers/GymDetailViewController.swift]: what if `equipmentList` is empty?

File: Uplift/Controllers/GymDetailViewController.swift [hunk 4]
```diff
@@ -282,6 +287,31 @@ func getHoursHeight() -> CGFloat {
...
+        equipmentDictionary[category] = []
+
+        sorted.forEach { (equipment) in
+            if equipment.equipmentType == category {
+                equipmentDictionary[category]?.append(equipment)
+            } else {
+                category = equipment.equipmentType
+                equipmentDictionary[category] = [equipment]
+            }
+        }
```
Comment (kevinchan159) [inline Uplift/Controllers/GymDetailViewController.swift]: the logic here seems to be more complicated than it has to be. essentially we want to build `equipmentDictionary` to be a map from a category to all equipments in that category. the current logic sorts the `equipmentList` by `equipmentType` and then we maintain the current `category` to build this dictionary. instead, we can simply just do:
```
let equipmentDictionary = equipmentList.reduce([:], { dict, equipment in 
  if equipment.equipmentType in dict {
    dict[equipment.equipmentType]?.append(equipment)
  } else {
    dict[equipment.equipmentType] = [equipment]
  }
}
```

File: Uplift/Controllers/GymDetailViewController.swift [hunk 5]
```diff
@@ -282,6 +287,31 @@ func getHoursHeight() -> CGFloat {
...
+
+        var equipmentCategories: [EquipmentCategory] = []
+        for (categoryName, equipment) in equipmentDictionary {
+            equipmentCategories.append(EquipmentCategory(categoryName: categoryName, equipment: equipment))
+        }
```
Comment (kevinchan159) [inline Uplift/Controllers/GymDetailViewController.swift]: `let equipmentCategories = equipmentDictionary.map({EquipmentCategory(categoryName: $0, equipment: $1 })`

File: Uplift/Models/Gym.swift [hunk 6]
```diff
@@ -59,6 +60,15 @@ struct Gym {
...
         gymHours = gymHoursList
+
+        var facilitiesList: [Facility] = []
+        let facilitiesData = gymData.facilities
+        facilitiesData.forEach({ (facility) in
+            if let facility = facility {
+                facilitiesList.append(Facility(gymData: facility))
+            }
+        })
+        facilities = facilitiesList
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: can simplify to
```
facilities = gymData.facilities.compactMap({ facility in
  if let facility = facility { 
    return Facility(gymData: facility) 
  }
  return nil
})
```

File: Uplift/Models/Gym.swift [hunk 7]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
             dayOfWeek = 0
         }
     }
+
+    init(gymHoursDataId: AllGymsQuery.Data.Gym.Facility.Time?) {
+        if let gymHoursData = gymHoursDataId {
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: why is it called `gymHoursDataId`? why not `facilityTime`?

File: Uplift/Models/Gym.swift [hunk 8]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+        } else {
+            openTime = Date()
+            closeTime = openTime
+            dayOfWeek = 0
+        }
+    }
+}
+
+struct Facility {
+    var equipment: [Equipment]
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: newline after 156

File: Uplift/Models/Gym.swift [hunk 9]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+    var miscInformation: [String]
+    var name: String
+    var times: [DailyGymHours]
+
+    init(gymData: AllGymsQuery.Data.Gym.Facility) {
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: isnt this `facilityData`?

File: Uplift/Models/Gym.swift [hunk 10]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+        if let equipmentGymData = gymData.equipment {
+            equipment = equipmentGymData.compactMap({ (equipmentData) -> Equipment? in
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: delete parens around `equipmentData`

File: Uplift/Models/Gym.swift [hunk 11]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+            equipment = []
+        }
+
+        var miscInformationList: [String] = []
+        if let informationData = gymData.miscInformation {
+            informationData.forEach({ (info) in
+                if let info = info {
+                    miscInformationList.append(info)
+                }
+            })
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]:
```
miscInformationList = informationData.compactMap({ ... })
```

File: Uplift/Models/Gym.swift [hunk 12]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+        }
+        miscInformation = miscInformationList
+
+        name = gymData.name
+
+        times = gymData.times.compactMap({ (gymHoursDataId) -> DailyGymHours? in
+            guard let gymHoursDataId = gymHoursDataId else { return nil }
+            return DailyGymHours(gymHoursDataId: gymHoursDataId)
+        })
+    }
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: newline after 189

File: Uplift/Models/Gym.swift [hunk 13]
```diff
@@ -127,4 +139,72 @@ struct DailyGymHours {
...
+}
+
+struct Equipment {
```
Comment (kevinchan159) [inline Uplift/Models/Gym.swift]: newline at beginning and end of struct

File: Uplift/Views/GymDetail/Facilities/EquipmentListCell.swift [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    override func configure(for models: [EquipmentCategory]) {
+        super.configure(for: models)
+
+        reloadLayout()
+    }
+
+    // When integrated:
+    // sizeForItemAt: 
+    // let height = EquipmentListCell.getHeight(models: equipment)
+    // return CGSize(width: width, height: height + 24.0)
```
Comment (kevinchan159) [inline Uplift/Views/GymDetail/Facilities/EquipmentListCell.swift]: what is this for?

File: Uplift/Views/GymDetail/Facilities/EquipmentListItemCell.swift [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    override func configure(for equipmentCategory: EquipmentCategory) {
+        super.configure(for: equipmentCategory)
+
+        titleLabel.text = equipmentCategory.categoryName
+        equipment = equipmentCategory.equipment
+        equipmentList.attributedText = generateListString(for: equipment.map { $0.name })
+        quantityList.attributedText = generateListString(for: equipment.map { $0.quantity }, alignment: .right)
+        quantityList.sizeToFit()
+
+        equipmentListTrailingConstraint?.update(offset: -listsSpacing)
```
Comment (kevinchan159) [inline Uplift/Views/GymDetail/Facilities/EquipmentListItemCell.swift]: why do we need this `trailingConstraint`? it doesn't seem like `listsSpacing` ever gets changed

File: Uplift/Controllers/GymDetailViewController.swift (hunk 1 above)
Comment (ywsang) [reply Uplift/Controllers/GymDetailViewController.swift]: only all uppercased

File: Uplift/Views/GymDetail/Facilities/EquipmentListCell.swift (hunk 14 above)
Comment (ywsang) [reply Uplift/Views/GymDetail/Facilities/EquipmentListCell.swift]: for me to remember how to put it back in to test

**Pull request merged**